### PR TITLE
Get block REST API

### LINKED ISSUE #3577: Get Block REST API
### Problem

HIP-415 is codifying the concept of blocks in relation to the Hedera network. The mirror node should expose block information via its REST API for use by bridges and other clients.

### Solution

The new mirror node `/api/v1/blocks/{hashOrNumber}` REST API will return the block information associated with a specific block. The path parameter will accept either a block hash or block number to uniquely identify a specific block. The block hash can be the 64 characters or 96 characters form and with or without the `0x` prefix to indicate it is hex encoded. The JSON response will be the same as [/api/v1/blocks](#1933) but without the pagination structure:

```json
{
  "count": 4,
  "gas_used": 50000000,
  "hapi_version": "0.24.0",
  "hash": "0xa4ef824cd63a325586bfe1a66396424cd33499f895db2ce2292996e2fc5667a69d83a48f3883f2acab0edfb6bfeb23c4",
  "logs_bloom": "0x549358c4c2e573e02410ef7b5a5ffa5f36dd7398",
  "name": "2022-04-07T16_59_23.159846673Z.rcd",
  "number": 19533336,
  "previous_hash": "0x4fbcefec4d07c60364ac42286d5dd989bc09c57acc7370b46fa8860de4b8721e63a5ed46addf1564e4f8cd7b956a5afa",
  "size": 8489,
  "timestamp": {
    "from": "1649350763.159846673"
    "to": "1649350763.382130000"
  }
}
```

Query parameters: None

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by natanasow**
**Description**:

**Related issue(s)**:

Fixes #3577

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by natanasow** [comment]
@steven-sheehy Do we need some kind of validation for the input param /blocks/:hashOrNumber, and what's the expected behavior? If the user passes an invalid hash (different length) or invalid number (/api/v1/blocks/testinput), what should we return, an empty object or an error like that :point_down: 
```
{
    "_status": {
        "messages": [
            {
                "message": "Invalid parameter: hashOrNumber"
            }
        ]
    }
}
```

**Comment by Nana-EC** [comment]
> @steven-sheehy Do we need some kind of validation for the input param /blocks/:hashOrNumber, and what's the expected behavior? If the user passes an invalid hash (different length) or invalid number (/api/v1/blocks/testinput), what should we return, an empty object or an error like that 👇
> 
> ```
> {
>     "_status": {
>         "messages": [
>             {
>                 "message": "Invalid parameter: hashOrNumber"
>             }
>         ]
>     }
> }
> ```

@natanasow  Yes, you'll definitely want to validate it. Failure should throw an `InvalidArgumentError.forParams(constants.filterKeys.HASHORNUMBER)` like you've noted with 
See `validateContractIdParam` & `validateContractIdAndConsensusTimestampParam` in the `contractController.js` as an example.
You'll wanna add a utils.isValidEthereumHashOrNumber(hashOrNumber) that can be utilized here and in the `utility.filterValidityChecks()`

File: hedera-mirror-rest/api/v1/openapi.yml [hunk 1]
```diff
@@ -224,6 +224,45 @@ paths:
           $ref: '#/components/responses/InvalidParameterError'
       tags:
         - blocks
+  /api/v1/blocks:
```
Comment (steven-sheehy) [inline hedera-mirror-rest/api/v1/openapi.yml:227]: This block duplicates the one above

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/__tests__/specs/blocks-get-by-hash-or-number-01-number.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{
+  "description": "Get block getByHashOrNumber with number param",
+  "setup": {
+    "recordFiles": [
+      {
+        "index": 16,
+        "count": 3,
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/specs/blocks-get-by-hash-or-number-01-number.json:7]: can you change `index` and `count` in this spec to be > js Number.MAX_SAFE_INTEGER? need to verify now we can correctly handle large numbers.

File: hedera-mirror-rest/api/v1/openapi.yml [hunk 3]
```diff
@@ -1523,6 +1566,24 @@ components:
...
                     type: string
+    HederaHash:
+      type: string
+      description: A hedera hash format.
+      minLength: 96
+      maxLength: 98
+      pattern: '^(0x)?[A-Fa-f0-9]{96}$'
+    EthereumHash:
+      type: string
+      description: A hedera hash format.
```
Comment (xin-hedera) [inline hedera-mirror-rest/api/v1/openapi.yml:1577]: incorrect description

File: hedera-mirror-rest/api/v1/openapi.yml [hunk 4]
```diff
@@ -1523,6 +1566,24 @@ components:
...
+      minLength: 64
+      maxLength: 66
+      pattern: '^(0x)?[A-Fa-f0-9]{64}$'
+    PositiveNumber:
+      type: string
+      description: A hedera hash format.
```
Comment (xin-hedera) [inline hedera-mirror-rest/api/v1/openapi.yml:1583]: incorrect description

File: hedera-mirror-rest/utils.js [hunk 5]
```diff
@@ -1190,6 +1190,20 @@ const isRegexMatch = (regex, value) => {
   return regex.test(value.trim());
 };
 
+const isRegexMatch = (regex, value) => {
+  return regex.test(value.trim());
+};
+
```
Comment (xin-hedera) [inline hedera-mirror-rest/utils.js:1196]: duplicate
```suggestion
```

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/service/recordFileService.js [hunk 6]
```diff
@@ -22,8 +22,10 @@
 
 const _ = require('lodash');
 
+const utils = require('../utils');
 const BaseService = require('./baseService');
 const {RecordFile} = require('../model');
+const {BlockViewModel} = require('../viewmodel');
```
Comment (Nana-EC) [inline hedera-mirror-rest/service/recordFileService.js:28]: Remove as not being used. Plus viewmodel should only be called in controller. @

**Review by Nana-EC — CHANGES_REQUESTED** [review]
One item, will circle back for 2nd review.
Agree with most comments

**Review by xin-hedera — APPROVED** [review]
LGTM

File: hedera-mirror-rest/service/recordFileService.js [hunk 7]
```diff
@@ -128,6 +133,31 @@ class RecordFileService extends BaseService {
 
     return rows.map((recordFile) => new RecordFile(recordFile));
   }
+
+  async getByHashOrNumber(hash, number) {
+    let whereStatement = '';
+    const params = [];
+    if (hash) {
+      const hashWithPrefix = utils.addHexPrefix(hash);
+      const hashWithoutPrefix = hashWithPrefix.substring(2);
```
Comment (Nana-EC) [inline hedera-mirror-rest/service/recordFileService.js:142]: Is `hash` a string already? If so this line isn't needed and you can just use `hash` below

File: hedera-mirror-rest/service/recordFileService.js [hunk 8]
```diff
@@ -128,6 +133,31 @@ class RecordFileService extends BaseService {
 
     return rows.map((recordFile) => new RecordFile(recordFile));
   }
+
+  async getByHashOrNumber(hash, number) {
+    let whereStatement = '';
+    const params = [];
+    if (hash) {
```
Comment (Nana-EC) [inline hedera-mirror-rest/service/recordFileService.js:140]: Move this data type conversion to the controller after param validation.

File: hedera-mirror-rest/service/recordFileService.js [hunk 9]
```diff
@@ -128,6 +133,31 @@ class RecordFileService extends BaseService {
...
+      const hashWithPrefix = utils.addHexPrefix(hash);
```
Comment (Nana-EC) [inline hedera-mirror-rest/service/recordFileService.js:141]: Why add prefix just to remove it again?
If the goal is just to remove the `0x` if it exists then we can do this with 
```suggestion
      const hashWithoutPrefix = hash.replace('0x', '');
```

File: hedera-mirror-rest/controllers/blockController.js [hunk 10]
```diff
@@ -28,12 +28,26 @@ const {RecordFileService} = require('../service');
...
 const constants = require('../constants');
+const {NotFoundError} = require('../errors/notFoundError');
+const {InvalidArgumentError} = require('../errors/invalidArgumentError');
 const {
   response: {
     limit: {default: defaultLimit, max: maxLimit},
   },
 } = require('../config');
 
+const validateHashOrNumber = (hashOrNumber) => {
```
Comment (Nana-EC) [inline hedera-mirror-rest/controllers/blockController.js:39]: conversion pruning of `0x` in hash should be done in here so service gets it as a useable param

File: hedera-mirror-rest/api/v1/openapi.yml [hunk 11]
```diff
@@ -1583,6 +1606,24 @@ components:
...
                     type: string
+    HederaHash:
+      type: string
+      description: A hedera hash format.
+      minLength: 96
+      maxLength: 98
+      pattern: '^(0x)?[A-Fa-f0-9]{96}$'
+    EthereumHash:
+      type: string
+      description: A ethereum hash format.
```
Comment (Nana-EC) [inline hedera-mirror-rest/api/v1/openapi.yml:1617]:
```suggestion
      description: A hex encoded ethereum transaction hash.
```

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good, just one more push with some items around hash format handling

File: hedera-mirror-rest/viewmodel/blockViewModel.js [hunk 12]
```diff
@@ -32,11 +32,11 @@ class BlockViewModel {
    * @param {Block} recordFile
    */
   constructor(recordFile) {
-    this.count = recordFile.count;
+    this.count = BigInt(recordFile.count);
     this.hapi_version = recordFile.getFullHapiVersion();
     this.hash = utils.addHexPrefix(recordFile.hash);
     this.name = recordFile.name;
-    this.number = recordFile.index;
+    this.number = BigInt(recordFile.index);
```
Comment (steven-sheehy) [inline hedera-mirror-rest/viewmodel/blockViewModel.js:39]: The fields are already BigInt when returned by the database.
```suggestion
    this.number =recordFile.index;
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/api/v1/openapi.yml [hunk 13]
```diff
@@ -224,6 +224,26 @@ paths:
...
+      operationId: getByHashOrNumber
+      parameters:
+        - $ref: '#/components/parameters/hashOrNumberPathParam'
+      responses:
+        200:
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/BlockResponse'
```
Comment (steven-sheehy) [inline hedera-mirror-rest/api/v1/openapi.yml:240]: Can remove useless `BlockResponse` wrapper and directly use '#/components/schemas/Block'

**Review by Nana-EC — APPROVED** [review]
LGTM.
Agree with Steven's last comments then we can merge

**Review by xin-hedera — DISMISSED** [review]

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**